README.md
<h2 align=center>j q u e</h2>
<p align=center>Query JSON in memory as though it were a Mongo database</p>
<p align=center><a href="https://pypi.org/project/jque/"><img src="https://img.shields.io/pypi/v/jque.svg" /></a> <a href="https://circleci.com/gh/j6k4m8/jque"><img src="https://img.shields.io/circleci/project/github/RedSparr0w/node-csgo-parser.svg" /></a> <img src="https://img.shields.io/badge/extremely_rad-%F0%9F%A4%99-blue.svg" /> <img src="https://img.shields.io/github/license/j6k4m8/jque.svg" /><a href="https://codecov.io/gh/j6k4m8/jque">
  <img src="https://codecov.io/gh/j6k4m8/jque/branch/master/graph/badge.svg" />
</a></p>

## Installation

```shell
pip3 install jque
```

## Usage

```python
import jque
```

`jque` accepts a variety of inputs to the constructor.

Pass a list of dicts:
```python
data = jque.jque([
    { "name": "john" }, 
    { "name": "paul" }, 
    { "name": "george" }, 
    { "name": "ringo" }
])
```

Pass a JSON filename:
```python
DATAFILE = "~/my/big/data.json"
data = jque.jque(DATAFILE)
```

Now you can query this dataset using Mongo-like syntax:

```python
>>> data.query({ "name": {"$neq": "paul"} })
[
    { "name": "john" },
    { "name": "george" }, 
    { "name": "ringo" }
]
```

### Arguments to `query`:

| Arg | Description |
|-----|-------------|
| `wrap` (`boolean` : `True`) | Whether to wrap the resultant dataset in a new `jque` object. This allows chaining, like `jque.query(...).query(...)`, if you're the sort of person to do that. Pass `False` to get back a `list` instead. |


### Another example!

```python
data = jque.jque([{
    "_id": "ABC",
    "name": "Arthur Dent",
    "age": 42,
    "current_planet": "earth"
}, {
    "_id": "DE2",
    "name": "Penny Lane",
    "age": 19,
    "current_planet": "earth"
}, {
    "_id": "123",
    "name": "Ford Prefect",
    "age": 240,
    "current_planet": "Brontitall"
}])
```

```python
teenage_earthlings = data.query({
    "current_planet": {"$eq": "earth"},
    "age": { "$lte": 20, "$gte": 10 }
})
```

Which returns:

```python
[{
    "_id": "DE2",
    "name": "Penny Lane",
    "age": 19,
    "current_planet": "earth"
}]
```


Use Python lambdas as a filter:

```python
>>> libraries = jque.jque([
...     {"name": "jque", "language": "Python"}, 
...     {"name": "react", "language": "node"}
... ])
>>> list(libraries.query({
...     'language': lambda x: x[:2] == "Py"
... }))
[{"name": "jque", "language": "Python"}]
```



CHANGELOG.md
# Changelog

- **0.1.3** (August 3, 2019)
    - Add `parallel` argument to constructor to toggle parallelism (using [`deco`](https://github.com/alex-sherman/deco))
    - Add `limit` argument to queries to stop search


run_tests.sh
pytest
pylint ./jque


tests/test_jque.py
#!/usr/bin/env python3

import os
import unittest

import jque

USERS = open(os.path.dirname(__file__) + "/data/users.json").read()

SMALL = [{
    "_id": "ABC",
    "name": "Arthur Dent",
    "age": 42,
    "current_planet": "earth"
}, {
    "_id": "DE2",
    "name": "Penny Lane",
    "age": 19,
    "current_planet": "earth"
}, {
    "_id": "123",
    "name": "Ford Prefect",
    "age": 240,
    "current_planet": "Brontitall"
}]


class TestJque(unittest.TestCase):

    def test_initialize(self):
        self.assertIsNotNone(jque.jque(USERS).data)

    def test_basic_filter(self):
        self.assertEqual(
            jque.jque(SMALL).query(
                {"name": {"$eq": "Ford Prefect"}})[0]['_id'],
            "123"
        )

    def test_literal_filter(self):
        self.assertEqual(
            jque.jque(SMALL).query({"name": "Ford Prefect"}).data[0]['_id'],
            "123"
        )

    def test_multi_literal_filter(self):
        self.assertEqual(
            jque.jque(USERS).query({
                "name": "Glenna Reichert",
                "username": "Delphine",
            }).data[0]['id'],
            9
        )

    def test_multi_filter(self):
        self.assertEqual(
            jque.jque(USERS).query({
                "username": "Delphine",
                "id": {"$gt": 9}
            }).data[0]['id'],
            109
        )

    def test_can_index(self):
        self.assertEqual(
            jque.jque(USERS).data[0],
            jque.jque(USERS)[0]
        )

    def test_length(self):
        self.assertEqual(
            len(jque.jque(USERS).data),
            len(jque.jque(USERS))
        )

    def test_accepts_filename(self):
        self.assertEqual(
            len(jque.jque(os.path.dirname(__file__) + "/data/users.json").data),
            len(jque.jque(USERS).data)
        )

    def test_list_in(self):
        self.assertEqual(
            len(jque.jque(USERS).query({
                "id": {
                    "$in": [9, 109]
                }
            })),
            2
        )

    def test_list_nin(self):
        self.assertEqual(
            len(jque.jque(USERS).query({
                "id": {
                    "$nin": [9, 109]
                }
            })),
            len(jque.jque(USERS)) - 2
        )

    def test_lambda(self):
        self.assertEqual(
            len(jque.jque(USERS).query({
                "id": lambda x: x not in [9, 109, 10]
            })),
            len(jque.jque(USERS)) - 3
        )

    def test_limit(self):
        data = jque.jque([{
            "_id": "ABC",
            "name": "Arthur Dent",
            "age": 42,
            "current_planet": "earth"
        }, {
            "_id": "DE2",
            "name": "Penny Lane",
            "age": 19,
            "current_planet": "earth"
        }, {
            "_id": "123",
            "name": "Ford Prefect",
            "age": 240,
            "current_planet": "Brontitall"
        }], parallel=True)
        self.assertEqual(
            len(data.query({"current_planet": "earth"})), 2
        )
        self.assertEqual(
            len(data.query({"current_planet": "earth"}, limit=1)), 1
        )


jque/__init__.py
#!/usr/bin/env python3

"""
In-Memory Mongo-Flavored Queries.

jque is a Python module that lets you query in-memory lists of dicts as though
they were in a Mongo database.
"""
from typing import List

import json
import copy
import types
from deco import synchronized, concurrent

__version__ = "0.1.3"


_OPERATORS = {
    "$eq": lambda x, y: x == y,
    "$neq": lambda x, y: x != y,
    "$lt": lambda x, y: x < y,
    "$lte": lambda x, y: x <= y,
    "$gt": lambda x, y: x > y,
    "$gte": lambda x, y: x >= y,
    "$in": lambda x, y: x in y,
    "$nin": lambda x, y: x not in y,
}


@concurrent
def _check_record_parallel(qr, record):
    return _check_record(qr, record)


def _check_record(qr, record):
    for key, qual in qr.items():
        if isinstance(qual, dict):
            for op, val in qual.items():
                if op not in _OPERATORS:
                    raise ValueError(
                        "'{}' is not a valid operator.".format(op)
                    )
                if not _OPERATORS[op](record[key], val):
                    return False
        elif isinstance(qual, types.FunctionType):
            if not qual(record[key]):
                return False
        else:
            if record[key] != qual:
                return False
    return True


class jque:
    """
    A JSON query class that subsets the behavior of MongoDB queries.

    Uses $-notation for mongo operators.

        data = jque.jsonque([{
            "_id": "ABC",
            "name": "Arthur Dent",
            "age": 42,
            "current_planet": "earth"
        }, {
            "_id": "DE2",
            "name": "Penny Lane",
            "age": 19,
            "current_planet": "earth"
        }, {
            "_id": "123",
            "name": "Ford Prefect",
            "age": 240,
            "current_planet": "Brontitall"
        }])

        teenage_earthlings = data.query({
            "current_planet": "earth",
            "age": { "$lte": 20, "$gte": 10 }
        })

    """

    OPERATORS = _OPERATORS

    def __init__(self, data, parallel=False):
        """
        Create a new jque object. Pass `data`, which must be a
        string or a list. If a list, each item should be a dictionary.
        If a string, it can either be a JSON string (from Python's
        json.dumps or JS's JSON.stringify) or a filename that points
        to a .json file on disk.
        """
        self.parallel = parallel
        if isinstance(data, str):
            try:
                data = json.loads(data)
            except ValueError:
                data = json.loads(open(data, 'r').read())
        elif not isinstance(data, list):
            raise ValueError("'data' argument must be a string or a list.")

        self.data = data

    def __getitem__(self, key):
        return self.data[key]

    def __len__(self):
        return len(self.data)

    def query(self, qr, wrap=True, limit=None):
        """
        Query the records for a desired trait.

        Arguments:
            qr (dict): a dict where all keys are included in all records.
            wrap (bool : True): If the result should be rewrapped in a
                new jque object.

        Examples:

        >>> data = jque([{
        ...     "_id": "ABC",
        ...     "name": "Arthur Dent",
        ...     "age": 42,
        ...     "current_planet": "earth"
        ... }, {
        ...     "_id": "DE2",
        ...     "name": "Penny Lane",
        ...     "age": 19,
        ...     "current_planet": "earth"
        ... }, {
        ...     "_id": "123",
        ...     "name": "Ford Prefect",
        ...     "age": 240,
        ...     "current_planet": "Brontitall"
        ... }])
        >>> len(data.query({ current_planet: "earth" })) == 2
        True
        >>> len(data.query({ current_planet: "earth" }), limit=1) == 1
        True
        """

        if self.parallel:
            # TODO: Concurrent support for limits
            results = _parallel_query(qr, self.data)
            filtered_results = [
                d for d, r in zip(self.data, results) if r
            ][:limit]
            if wrap:
                return jque(filtered_results)
            return filtered_results

        filtered_data = []
        for record in self.data:
            include = _check_record(qr, record)
            if include:
                filtered_data.append(record)
                if limit and len(filtered_data) >= limit:
                    break
        if wrap:
            return jque(filtered_data)
        return filtered_data


@synchronized
def _parallel_query(qr, data) -> List[bool]:
    results = [False] * len(data)
    for i in range(len(data)):
        results[i] = _check_record(qr, data[i])
    return results
